### PR TITLE
[SYCL][ESIMD][EMU] ESIMD_CPU Kernel launch and ESIMD_EMU backend loading

### PR DESCRIPTION
**Pull request by dongkyunahn-intel**
* This PR is for enabling kernel launching for ESIMD_CPU

* Also contains emulated intrinsics for memory operations

* esimd_cpu backend is loaded in SYCL runtime

* Base PR : https://github.com/intel/llvm/pull/4011

### CODE REVIEW

**Review by smaslov-intel — DISMISSED** [review]
PI-related changes LGTM

**Comment by dongkyunahn-intel** [comment]
I'll revert some changes (relocation + update) in following files and apply them in PR#4011 due to changes in Device interface.

- esimd_emu_functions_v1.h
- esimdcpu_device_interface.hpp

**Comment by dongkyunahn-intel** [comment]
Revert is complete.

**Review by smaslov-intel — DISMISSED** [review]
PI related changes LGTM

**Comment by dongkyunahn-intel** [comment]
Jenkins and buildbot jobs are failing due to rebase failure. I'll make single commit using these change sets and merge it to the `origin/sycl` branch after PR#4011 is merged to `origin/sycl`.

File: sycl/include/CL/sycl/accessor.hpp
Comment (kbobrovs) [inline sycl/include/CL/sycl/accessor.hpp:675]: we can't add public API to std SYCL classes. can use the proxy approach I used for similar purpose. @romanovvlad - what would be best here?

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (romanovvlad) [inline sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:520]: It seem this namespace declared in the "globally"(not in the namespace user isn't supposed to use).  If so, suggest moving it to `cl::sycl::__detail` namespace.

**Review by romanovvlad — COMMENTED** [review]
LGTM except some "non-functional" comments".
(haven't reviewed logic of __esimd* builtins)

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:518]: I wanted to put every code and header file inclusion required for emulation support in one place enclosed by `#ifndef __SYCL_DEVICE_ONLY__`/`#endif`. It's okay to move these header file inclusions in the beginning of the file.

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:520]: Will move to suggested name space.

File: sycl/include/CL/sycl/accessor.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/accessor.hpp:675]: Will fix it to use `AccessorPrivateProxy`.

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:538]: `raw_send` namespace is put under `sycl::ext::intel::experimental::esimd::detail` (aliased with `__SEIEED`)

File: sycl/include/CL/sycl/accessor.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/accessor.hpp]: `get_pointer()` added in `accessor.hpp` is now removed. `AccessorPrivateProxy` is used for accessing protected `getPtr()` function.

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp]: Applied.

**Review by romanovvlad — APPROVED** [review]
LGTM

**Review by kbobrovs — APPROVED** [review]

**Comment by dongkyunahn-intel** [comment]
Buildbot job failed because of missing implementation of PI_API. PR containing PI_API implementation is under review (https://github.com/intel/llvm/pull/4011). I'll try to merge this change set after the PR is merged into the SYCL branch.

**Review by kychendev — DISMISSED** [review]

**Comment by dongkyunahn-intel** [comment]
I fixed a problem from `void` kernel function argument, but buildbot failed as the PR#4011 that this PR depends on was reverted from `origin/sycl` due to post-commit failure. I'll try to merge/rebase after same change set is merged again (https://github.com/intel/llvm/pull/4430)

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/atomic_intrin.hpp
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/detail/atomic_intrin.hpp:12]:
```suggestion
// This function implements atomic update of pre-existing variable in the absense
// of C++ 20's atomic_ref. 
template <typename Ty> Ty atomic_add_fetch(Ty *ptr, Ty val) {
```

File: sycl/include/CL/sycl/handler.hpp
Comment (kbobrovs) [inline sycl/include/CL/sycl/handler.hpp:724]: IIRC, KI::isESIMD() was used to adjust extraction of actual pointer data from accessor when setting kernel arguments. Will it still work correctly when isESIMD is defaulted to false with this change?

File: sycl/include/CL/sycl/handler.hpp
Comment (kbobrovs) [inline sycl/include/CL/sycl/handler.hpp:692]: This version is supposed to be used with parallel_for_workgroup, which is not supported for ESIMD. So I'd make operator() just throw, and comment out `detail::runKernelWithArg(MKernelFunc, Arg.get_group());`

Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp:692]: Following tests could fail without this change even if esimd_cpu build is disable by default. This is because this file is used for all backends and host kernel. I applied this change because of them.

https://github.com/intel/llvm-test-suite/blob/intel/SYCL/Basic/stream/stream.cpp
https://github.com/intel/llvm-test-suite/blob/intel/SYCL/SpecConstants/2020/kernel_lambda_with_kernel_handler_arg.cpp

Comment (kbobrovs) [reply sycl/include/CL/sycl/handler.hpp:692]: Tests containing parallel_for_work_group should not be run with ESIMD, as they will fail. You may exclude those tests for ESIMD_CPU_EMU. @vladimirlaz, can we create a lit feature for ESIMD CPU EMU plugin similar to opencl and level zero?

Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp:692]: I'll revert changes related to this 'group' argument support after a lit feature for ESIMD CPU Emulation plug-in is created.

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:1523]: I'll replace implementations for `raw_send` support with exception throwing now. When unit tests are ready, I'll bring them back.

File: sycl/include/CL/sycl/handler.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp:727]: `isESIMD()` argument is recovered.

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:39]: this line leads to the failure of esimd/odr.cpp test failure. `esimdcpu_device_interface.hpp` defines the `getESIMDDeviceInterface` function, and is included into two sources which are then linked.

You should either add `inline` to this function or move its definition to some .cpp file which gets linked into libsycl.so. I'd prefer the second. @romanovvlad - do you have a preference?

File: sycl/include/CL/sycl/handler.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp:693]: > I'll revert changes related to this 'group' argument support after a lit feature for ESIMD CPU Emulation plug-in is created.

Am I right here? Should `ResetHostKernel` also have to cover `sycl::group<>` argument type because `MHostKernel.reset()` from left side is changed with call to `ResetHostKernel`?

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp]: Implementation removed.

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:39]: @romanovvlad , what is your preference?

And, if a new `.cpp` files is to be created, where should it be placed? `$src/llvm/sycl/source/detail`? (where `pi.cpp` is located)

File: sycl/include/CL/sycl/handler.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp:693]: @vladimirlaz / @kbobrovs , are we going to create a lit feature for ESIMD_CPU emulation for this type of exclusion?

And, how did tests with `parallel_for_work_groups` pass for ESIMD so far?

Comment (vladimirlaz) [reply sycl/include/CL/sycl/handler.hpp:693]: I would like to highlight that on-device tests (the tests which actually executed on target BE) have been removed from the intel/llvm repo to intel/llvm-test-suite (see https://github.com/intel/llvm/pull/4393).
So, these test should be migrated to intel/llvm-test-suite. Correspondign feature can be added the (see HIP BE as an example).

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (romanovvlad) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:39]: The second option sounds better.

Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:39]: Location & name of the new cpp file : `sycl/source/esimdcpu_device_interface.cpp`

(REF : Header `sycl/include/sycl/ext/intel/experimental/esimd/emu/detail/esimdcpu_device_interface.hpp`)

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [inline sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:725]: This change causes failures from following test with 'host' device
- SYCL/ESIMD/histogram.cpp
- SYCL/ESIMD/histogram_2d.cpp
- SYCL/ESIMD/histogram_raw_send.cpp
- SYCL/ESIMD/linear/linear.cpp
- SYCL/ESIMD/vadd_2d.cpp

This is because with 'host' device loaded, `getPluginOpaqueData<backend::esimd_cpu>()` inside of `getESIMDDeviceInterface` should fail as `getPluginOpaqueData` calls `getPlugin<backend::esimd_cpu>()` (Failure code : 'pi::getPlugin couldn't find plugin -59').

Unless there is a way to distinguish between 'host' type and 'esimd_cpu' type at runtime within this intrinsic implementation, I think these failing samples must be marked as UNSUPPORTED for host device type.

File: sycl/include/CL/sycl/handler.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp:693]: @vladimirlaz, the tests I mentioned above are already in intel/llvm-test-suite. You can see in their URLs.

Comment (vladimirlaz) [reply sycl/include/CL/sycl/handler.hpp:693]: @dongkyunahn-intel you can create LIT feature esimd_cpu and use it for the tests (like @kbobrovs suggested).

File: sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:725]: @kbobrovs , Can I mark host-device unsupported for above ESIMD tests in intel/llvm-test-suite?

Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:725]: PR created for marking UNSUPPORTED for failing ones : https://github.com/intel/llvm-test-suite/pull/492

Comment (kbobrovs) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:725]: I wonder how other tests pass then, like vadd_1d? Isn't the PI call sequence the same for them?

Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/detail/memory_intrin.hpp:725]: Other passing tests like `vadd_1d` either do not have `HOST_RUN_PLACEHOLDER` line. (https://github.com/intel/llvm-test-suite/blob/1a770de1b40f9cce67b63d0e42ac5d06984f7ce3/SYCL/ESIMD/vadd_1d.cpp#L11)

My changes in this file are applied only for `#ifndef SYCL_DEVICE_ONLY` therefore only host mode execution is impacted - irrelevant to PI call sequence. (ESIMD_CPU build is disabled by default in origin/sycl)

There are 17 samples with the `HOST_RUN_PLACEHOLDER` line from `SYCL/ESIMD/` directory.
```SYCL/ESIMD/BitonicSortKv2.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/PrefixSum.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out 20
SYCL/ESIMD/Prefix_Local_sum1.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out 20
SYCL/ESIMD/Prefix_Local_sum2.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out 20
SYCL/ESIMD/Prefix_Local_sum3.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/Stencil.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/histogram.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/histogram_2d.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/histogram_raw_send.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/kmeans/kmeans.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out %S/points.csv
SYCL/ESIMD/linear/linear.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out %S/linear_in.bmp %S/linear_gold_hw.bmp
SYCL/ESIMD/matrix_transpose_usm.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/reduction.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/stencil2.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/test_id_3d.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/vadd_2d.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
SYCL/ESIMD/vadd_usm.cpp:// RUN: %HOST_RUN_PLACEHOLDER %t.out
```
Among these 17 samples, the above five samples are impacted by this change set - all of them due to `media_block_load/store`

File: sycl/include/CL/sycl/handler.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp]: @vladimirlaz , what file should I update for creating a LIT feature?

**Comment by dongkyunahn-intel** [comment]
Memory intrinsic implementations are posted in https://github.com/intel/llvm/pull/4748.

**Comment by dongkyunahn-intel** [comment]
This PR has to be merged after PR#4728(https://github.com/intel/llvm/pull/4728) with renaming for `ESIMD_EMU`.

**Comment by dongkyunahn-intel** [comment]
@kbobrovs , @againull , @romanovvlad , I had to update the PR for missing changes for esimd_cpu/emulator renaming. Please review and approve.

**Review by kbobrovs — DISMISSED** [review]

**Review by againull — DISMISSED** [review]

**Comment by dongkyunahn-intel** [comment]
@romanovvlad , @v-klochkov , would you review/approve this PR?

**Comment by dm-vodopyanov** [comment]
@dongkyunahn-intel, can you please resolve the CI failures in Jenkins/Summary?

**Comment by dongkyunahn-intel** [comment]
@dm-vodopyanov ,

I've been trying to resolve the failure, but is not successful yet.

The failure seems to be segmentation fault (memory access violation) under windows environment and occurs only for one precision - `system::double16` (https://github.com/KhronosGroup/SYCL-CTS/blob/65856bfb5844fc73f1c4e0841bf4be4c6028eef9/tests/handler/handler_copy.cpp#L948). Other precisions like `int` and `double` are fine.

**Comment by dongkyunahn-intel** [comment]
It is found that the segmentation fault is because of violation against 16-byte memory access alignment required by XMM access. Currently, only warning message is printed out during compilation for cases where the requirement cannot be satisfied - https://github.com/intel/llvm/blob/6761e738c69ee4f5ca976298e43e3d023b02563d/sycl/include/CL/sycl/types.hpp#L545.

This issue has to be addressed in a separate PR.

**Comment by dongkyunahn-intel** [comment]
PR for memory access alignment is posted : https://github.com/intel/llvm/pull/4953

File: sycl/include/CL/sycl/detail/cg_types.hpp
Comment (s-kanaev) [inline sycl/include/CL/sycl/detail/cg_types.hpp:250]: This doesn't seem to be backward compatible change.
Please, add getter/setter methods or declare `friend` class or method.

File: sycl/include/CL/sycl/handler.hpp
Comment (s-kanaev) [inline sycl/include/CL/sycl/handler.hpp:1437]: Is this change really needed?

File: sycl/include/CL/sycl/detail/cg_types.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/detail/cg_types.hpp]: `friend` is added for `sycl::handler`. `MKernel` is brought back to `private`.

File: sycl/include/CL/sycl/handler.hpp
Comment (dongkyunahn-intel) [reply sycl/include/CL/sycl/handler.hpp]: Change reverted.

**Comment by dongkyunahn-intel** [comment]
@romanovvlad , @kbobrovs , @s-kanaev , please review and approve this PR.

Cause of failures from 'test_handler' is identified, and corresponding ticket was created & shared internally.

**Comment by dongkyunahn-intel** [comment]
Second review/approval request : @romanovvlad , @kbobrovs , @s-kanaev.

**Review by romanovvlad — APPROVED** [review]

**Comment by romanovvlad** [comment]
@kbobrovs, @v-klochkov , @smaslov-intel Could you please review and approve if the patch is OK for you?

**Review by smaslov-intel — APPROVED** [review]
PI related changes LGTM

**Comment by dongkyunahn-intel** [comment]
@kbobrovs , @v-klochkov , Please review and approve if this patch looks okay to you.

**Review by kbobrovs — APPROVED** [review]

File: sycl/source/detail/device_filter.cpp
Comment (v-klochkov) [inline sycl/source/detail/device_filter.cpp:94]: Late comment: This line would be better as below:
`"{host,opencl,level_zero,cuda,hip,esimd_emulator,*}.\n"`

Comment (dongkyunahn-intel) [reply sycl/source/detail/device_filter.cpp:94]: Will apply this change as piggyback in another PR later.

**Pull request merged**
